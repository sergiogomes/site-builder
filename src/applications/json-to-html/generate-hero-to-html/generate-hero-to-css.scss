.hero {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  height: 100dvh;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;

  .container {
    margin: auto;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    background-color: #333333a3;
    border-radius: 8px;
    padding: 20px;
    
    @media (min-width: 768px) {
      gap: 20px;
      padding: 80px;
    }
  }

  h2 {
    font-size: 3rem;
    margin-bottom: 10px;
    color: white;
    text-align: center;
    
    @media (min-width: 768px) {
      font-size: 5rem;
    }
  }

  a {
    text-decoration: none;
    color: black;
    background-color: white;
    padding: 10px 20px;
    border-radius: 8px;
    margin: auto;
    font-size: 1.5rem;
  }
}
