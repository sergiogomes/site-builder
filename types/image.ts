export type imageProps = {
  src: string
  alt: string
}
