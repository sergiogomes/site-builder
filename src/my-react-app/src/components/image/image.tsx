import { useState } from 'react'

import { type ImageProps } from '../../types/image'

type ImageFormProps = {
  isSubComponent: boolean
  onChange: (inputName: string, value: string) => void
}

function Image({ isSubComponent, onChange }: ImageFormProps) {

  if (isSubComponent) {
    return (
      <div>
        <ImageContent isSubComponent={isSubComponent} onChange={onChange} />
      </div>
    )
  }

  return (
    <form>
      <ImageContent isSubComponent={isSubComponent} onChange={() => {}} />
    </form>
  )
}

const initialImageObj: ImageProps = {
  src: '',
  alt: '',
}

function ImageContent({ isSubComponent, onChange }: ImageFormProps) {

  const [imageObj, setImageObj] = useState<ImageProps>(initialImageObj as ImageProps)

  const handleChange = (inputName: string, value: string) => {
    setImageObj((prev) => ({ ...prev, [inputName]: value }))
    if (isSubComponent) {
      onChange(inputName, value)
    }
  }

  return (
    <>
      <div className="form-field">
        <label htmlFor="image-src">Image Source</label>
        <input
          id="image-src"
          type="link" // Only link is supported for now
          placeholder="Image Source"
          value={imageObj.src}
          onChange={(e) => handleChange('src', e.target.value)}
        />
      </div>

      <div className="form-field">
        <label htmlFor="image-alt">Image Alt Text</label>
        <input
          id="image-alt"
          type="text"
          placeholder="Image Alt Text"
          value={imageObj.alt}
          onChange={(e) => handleChange('alt', e.target.value)}
        />
      </div>

      {!isSubComponent && (
        <button
          type="submit"
          onClick={() => console.log('Image Object:', imageObj)}
        >
          Salvar
        </button>
      )}

    </>
  )
}

export default Image
