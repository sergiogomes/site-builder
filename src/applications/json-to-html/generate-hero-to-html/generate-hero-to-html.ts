import { generateLinkToHtml } from '../generate-link-to-html/generate-link-to-html'
import './generate-hero-to-css.scss'

import { type ImageProps } from '../../../entities/image'
import { type LinkProps } from '../../../entities/link'

export type GenerateHeroToHtmlProps = {
  title?: string
  image?: ImageProps
  cta?: LinkProps
} & (
  | { title: string }
  | { image: ImageProps }
  | { cta: LinkProps }
)
export function generateHeroToHtml(props: GenerateHeroToHtmlProps): string {
  const { title, image, cta } = props

  const html: string[] = []

  const titleHtml = title ? `<h2>${title}</h2>` : ''
  const ctaHtml = cta ? generateLinkToHtml({ link: cta }) : ''

  if (titleHtml || image?.src || ctaHtml) {
    if (image?.src) {
      html.push(`<div class="hero" style="background-image: url('${image.src}');"><div class="container">`)
    } else {
      html.push('<div class="hero"><div class="container">')
    }
    html.push(titleHtml)
    html.push(ctaHtml)
    html.push('</div></div>')
  }

  return html.join('')
}
