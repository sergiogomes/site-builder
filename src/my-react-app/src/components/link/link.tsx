import { useState } from 'react'

import { type LinkProps } from '../../types/link'

type LinkFormProps = {
  isSubComponent: boolean
  onChange: (inputName: string, value: string) => void
}

function Link({ isSubComponent, onChange }: LinkFormProps) {

  if (isSubComponent) {
    return (
      <div>
        <LinkContent isSubComponent={isSubComponent} onChange={onChange} />
      </div>
    )
  }

  return (
    <form>
      <LinkContent isSubComponent={isSubComponent} onChange={() => {}} />
    </form>
  )
}

function LinkContent({ isSubComponent, onChange }: LinkFormProps) {
  const [linkObj, setLinkObj] = useState<LinkProps>({} as LinkProps)

  const handleChange = (inputName: string, value: string) => {
    setLinkObj((prev) => ({ ...prev, [inputName]: value }))
    if (isSubComponent) {
      onChange(inputName, value)
    }
  }

  return (
    <div>
      <div className="form-field">
        <label htmlFor="link-text">Link Text</label>
        <input
          id="link-text"
          type="text"
          placeholder="Link Text"
          value={linkObj.text}
          onChange={(e) => handleChange('text', e.target.value)}
        />
      </div>

      <div className="form-field">
        <label htmlFor="link-href">Link Href</label>
        <input
          id="link-href"
          type="link" // Only link is supported for now
          placeholder="Link Href"
          value={linkObj.href}
          onChange={(e) => handleChange('href', e.target.value)}
        />
      </div>

      <div className="form-field">
        <label htmlFor="link-anchor">Link Anchor</label>
        <input
          id="link-anchor"
          type="text"
          placeholder="Link Anchor"
          value={linkObj.anchor}
          onChange={(e) => handleChange('anchor', e.target.value)}
        />
      </div>

      <div className="form-field">
        <label htmlFor="link-target">Link Target</label>
        <input
          id="link-target"
          type="text"
          placeholder="Link Target"
          value={linkObj.target}
          onChange={(e) => handleChange('target', e.target.value)}
        />
      </div>

      {!isSubComponent && (
        <button
          type="submit"
          onClick={() => console.log('Link Object:', linkObj)}
        >
          Salvar
        </button>
      )}
    </div>
  )
}

export default Link
