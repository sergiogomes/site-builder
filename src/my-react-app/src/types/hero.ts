import { ImageProps } from "./image"
import { LinkProps } from "./link"

export type HeroProps = {
  title?: string
  image?: ImageProps
  cta?: LinkProps
} & (
  | { title: string }
  | { image: ImageProps }
  | { cta: LinkProps }
)
