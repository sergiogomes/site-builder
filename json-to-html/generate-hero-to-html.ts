import { imageProps } from '../types/image'
import { linkProps } from '../types/link'

type generateHeroToHtmlProps = {
  title: string
  image: imageProps
  cta: linkProps
}
export function generateHeroToHtml(props: generateHeroToHtmlProps): string {
  const { title, image, cta } = props

  return ''
}
