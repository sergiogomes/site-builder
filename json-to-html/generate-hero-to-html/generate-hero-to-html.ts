import { generateLinkToHtml } from '../generate-link-to-html/generate-link-to-html'
import { generateImageToHtml } from '../generate-image-to-html/generate-image-to-html'

import { type ImageProps } from '../../types/image'
import { type LinkProps } from '../../types/link'

export type GenerateHeroToHtmlProps = {
  title?: string
  image?: ImageProps
  cta?: LinkProps
} & (
  | { title: string }
  | { image: ImageProps }
  | { cta: LinkProps }
)
export function generateHeroToHtml(props: GenerateHeroToHtmlProps): string {
  const { title, image, cta } = props

  const html: string[] = []

  const titleHtml = title ? `<h2>${title}</h2>` : ''
  const imageHtml = image ? generateImageToHtml({ image }) : ''
  const ctaHtml = cta ? generateLinkToHtml({ link: cta }) : ''

  if (titleHtml || imageHtml || ctaHtml) {
    html.push('<div class="hero">')
    html.push(titleHtml)
    html.push(imageHtml)
    html.push(ctaHtml)
    html.push('</div>')
  }

  return html.join('')
}
