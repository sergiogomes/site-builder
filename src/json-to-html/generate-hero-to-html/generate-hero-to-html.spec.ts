import { compile } from 'sass-embedded'

import { generateHeroToHtml } from './generate-hero-to-html'

import { type ImageProps } from '../../../types/image'
import { type LinkProps } from '../../../types/link'


describe('generateHeroToHtml', () => {
  it('should generate HTML with title only', () => {
    const props = { title: 'Hero Title' }
    const result = generateHeroToHtml(props)
    
    expect(result).toMatchSnapshot()
  })

  it('should generate HTML with image only', () => {
    const image: ImageProps = { src: 'image.jpg', alt: 'An image' }
    const props = { image }
    const result = generateHeroToHtml(props)
    
    expect(result).toMatchSnapshot()
  })

  it('should generate HTML with cta only', () => {
    const cta: LinkProps = { href: 'https://example.com', text: 'Click here' }
    const props = { cta }
    const result = generateHeroToHtml(props)
    
    expect(result).toMatchSnapshot()
  })

  it('should generate HTML with title, image, and cta', () => {
    const image: ImageProps = { src: 'image.jpg', alt: 'An image' }
    const cta: LinkProps = { href: 'https://example.com', text: 'Click here' }
    const props = { title: 'Hero Title', image, cta }
    const result = generateHeroToHtml(props)
    
    expect(result).toMatchSnapshot()
  })

  it('should test scss', () => {
    const result = compile(__dirname + '/generate-hero-to-css.scss')

    expect(result.css).toMatchSnapshot()
  })
})
