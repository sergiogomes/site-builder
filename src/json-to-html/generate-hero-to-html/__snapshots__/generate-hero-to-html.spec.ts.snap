// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateHeroToHtml should compile Hero SCSS to CSS 1`] = `
".hero {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 20px;
  height: 100dvh;
  background-repeat: no-repeat;
  background-position: center;
  background-size: cover;
}
.hero .container {
  margin: auto;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: rgba(51, 51, 51, 0.6392156863);
  border-radius: 8px;
  padding: 20px;
}
@media (min-width: 768px) {
  .hero .container {
    gap: 20px;
    padding: 80px;
  }
}
.hero h2 {
  font-size: 3rem;
  margin-bottom: 10px;
  color: white;
  text-align: center;
}
@media (min-width: 768px) {
  .hero h2 {
    font-size: 5rem;
  }
}
.hero a {
  text-decoration: none;
  color: black;
  background-color: white;
  padding: 10px 20px;
  border-radius: 8px;
  margin: auto;
  font-size: 1.5rem;
}"
`;

exports[`generateHeroToHtml should generate HTML with cta only 1`] = `"<div class="hero"><div class="container"><a href="https://example.com">Click here</a></div></div>"`;

exports[`generateHeroToHtml should generate HTML with image only 1`] = `"<div class="hero" style="background-image: url('image.jpg');"><div class="container"></div></div>"`;

exports[`generateHeroToHtml should generate HTML with title only 1`] = `"<div class="hero"><div class="container"><h2>Hero Title</h2></div></div>"`;

exports[`generateHeroToHtml should generate HTML with title, image, and cta 1`] = `"<div class="hero" style="background-image: url('image.jpg');"><div class="container"><h2>Hero Title</h2><a href="https://example.com">Click here</a></div></div>"`;
